### DISCUSSION TITLE
What does "kfuncs have no UAPI guarantees" mean?

### DISCUSSION THREAD

**Discussion by Silvanoc** [isovalent/ebpf-docs#52] [Q&A] [answered]
There are at least two places in the documentation stating that "kfuncs have no UAPI guarantees":
1. In the [introduction](https://github.com/isovalent/ebpf-docs/blob/38ce77e51bc300c77707cfc1ef97962100b5bfd9/docs/linux/index.md?plain=1#L35): "KFuncs are not considered UAPI (User-space API) and are not subject to the same stability guarantees as the UAPI."
2. In the [concepts](https://github.com/isovalent/ebpf-docs/blob/38ce77e51bc300c77707cfc1ef97962100b5bfd9/docs/linux/concepts/kfuncs.md?plain=1#L11C33-L11C88): "Officially KFuncs are unstable, unlike helper functions, kfuncs have no UAPI guarantees."

IMHO these statements without further explanation result confusing.

When I first read it, I was like "The UAPI is (as the first reference also states) the User-Space API, but helper functions are not used in User-Space!". I knew that they belong to the declared BPF API and therefore they have higher guarantees than KFuncs. But I could not fully understand the relationship with the UAPI, and a search at the [UAPI Specifications](https://uapi-group.org/specifications/) was not yielding any result either.

Until realizing that the canonical helper functions declaration is in [include/uapi/linux/bpf.h](https://github.com/torvalds/linux/blob/67784a74e258a467225f0e68335df77acd67b7ab/include/uapi/linux/bpf.h#L1828)... That is what is meant by the implicit statement that "helper functions have UAPI guarantees", right?

If that is all right, then we might want to first document the UAPI guarantees on helper functions to put KFuncs in contrast. Otherwise IMHO the introduction page is already requiring a too high level of expertise.

I can try to propose something once I get an answer to the question.

**Comment by dylandreimerink** [comment, marked as answer]
Yes, that is all correct. 

> If that is all right, then we might want to first document the UAPI guarantees on helper functions to put KFuncs in contrast. Otherwise IMHO the introduction page is already requiring a too high level of expertise.

That is fair! lets do it

**Comment by Silvanoc** [comment]
Initial draft proposal: https://github.com/isovalent/ebpf-docs/pull/53
